File: web/src/features/teacher/course-completion-terms/model/CourseCompletionTerms.ts
export type CourseCompletionTerms = {
  id: number;
  course_id: number;
  teacher_user_id: number;
  organization_id: number | null;
  version: number;
  status: string;
  completion_reward_amount: string;
  max_enrolled_students: number;
  reward_policy_id: number | null;
  proposed_by_user_id: number;
  accepted_by_user_id: number | null;
  accepted_at: string | null;
  activated_at: string | null;
  superseded_at: string | null;
  created_at: string;
  updated_at: string;
};

export type CourseCompletionTermsAuditEvent = {
  id: number;
  terms_id: number;
  course_id: number;
  actor_user_id: number;
  event_type: string;
  previous_status: string | null;
  new_status: string;
  completion_reward_amount: string;
  max_enrolled_students: number;
  note: string | null;
  created_at: string;
};

export type CourseCompletionTermsHistory = {
  active_terms: CourseCompletionTerms | null;
  terms: CourseCompletionTerms[];
  audit_events: CourseCompletionTermsAuditEvent[];
};

File: src/infra/postgres/models/course_completion_terms.rs
use bigdecimal::BigDecimal;
use chrono::{DateTime, Utc};
use diesel::prelude::*;

use crate::infra::postgres::schema::course_completion_terms;

#[derive(Queryable, Identifiable, Selectable, Debug, Clone)]
#[diesel(table_name = course_completion_terms)]
pub struct CourseCompletionTerms {
    pub id: i64,
    pub course_id: i32,
    pub teacher_user_id: i32,
    pub organization_id: Option<i32>,
    pub version: i32,
    pub status: String,
    pub completion_reward_amount: BigDecimal,
    pub max_enrolled_students: i32,
    pub reward_policy_id: Option<i64>,
    pub proposed_by_user_id: i32,
    pub accepted_by_user_id: Option<i32>,
    pub accepted_at: Option<DateTime<Utc>>,
    pub activated_at: Option<DateTime<Utc>>,
    pub superseded_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Insertable, Debug)]
#[diesel(table_name = course_completion_terms)]
pub struct NewCourseCompletionTerms {
    pub course_id: i32,
    pub teacher_user_id: i32,
    pub organization_id: Option<i32>,
    pub version: i32,
    pub status: String,
    pub completion_reward_amount: BigDecimal,
    pub max_enrolled_students: i32,
    pub proposed_by_user_id: i32,
}
